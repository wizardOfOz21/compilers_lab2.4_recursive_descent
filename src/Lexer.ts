import { addGroups, alternativeCombine, match, RegexpMap } from "./regex_utils";
import { EOF, ERROR, Token } from "./token";
import { Position } from "./utils";

export type LexemMap = RegexpMap;

export const NEWLINE = "NEWLINE";
export const NEWLINE_R = "NEWLINE_R";
export const SPACE = "SPACE";

export const ignoreLexems: LexemMap = {
    SPACE: /[^\S\r\n]+/,
    COMMENT_1: /\{[^}]*\}/u,
    COMMENT_2: /\(\*((?!\*\))[\s\S])*\*\)/u,
}

const defaultLexems: LexemMap = {
    NEWLINE: /\n/,
    NEWLINE_R: /\r\n/,
};

class Lexer {
    private ignore: string[];
    private lexems: string[];
    private regex: RegExp;
    private row: number;
    private col: number;
    private error: boolean;
    private input: string;

    getRegex(lexems: LexemMap): LexemMap {
        return addGroups({...defaultLexems, ...lexems });
    }

    constructor(input: string, lexems: LexemMap) {
        this.input = input;
        this.lexems = {
            ...Object.keys(defaultLexems),
            ...Object.keys(lexems),
            ...Object.keys(ignoreLexems),
        };
        this.regex = alternativeCombine(this.getRegex({...ignoreLexems, ...lexems}), "u");
        this.row = 1;
        this.col = 1;
        this.error = false;
    }

    step(length: number) {
        this.input = this.input.slice(length);
    }

    parse(): Token {
        if (this.input == "") {
            return new Token(EOF, "", new Position(-1, -1));
        }
        let matchResult = match(this.input, this.regex);
        let value = matchResult && matchResult.input;
        let groups = matchResult && matchResult.groups;
        let lexem =
            groups &&
            Object.keys(groups).find(
                (key) => !!groups[key] || groups[key] == ""
            );

        if (!matchResult) {
            let col = this.col;
            this.step(1);
            this.col += 1;
            if (this.error) {
                return this.parse();
            } else {
                this.error = true;
                return new Token(ERROR, "", new Position(this.row, col));
            }
        }
        if (this.error) {
            this.error = false;
        }
        if (lexem == NEWLINE) {
            this.step(1);
            this.row += 1;
            this.col = 1;
            return this.parse();
        } else if (lexem == NEWLINE_R) {
            this.step(2);
            this.row += 1;
            this.col = 1;
            return this.parse();
        } else if (lexem in ignoreLexems) {
            value = groups[lexem];
            this.step(value.length);
            this.col += value.length;
            return this.parse();
        }
        value = groups[lexem];
        this.step(value.length);
        let col = this.col;
        this.col += value.length;

        return new Token(lexem, value, new Position(this.row, col));
    }
}

export default Lexer;

export function printLexems(lexer: Lexer) {
    let token: Token = lexer.parse();
    while (!token.isEof()) {
        if (token.isError()) {
            console.log(`syntax error (${token.pos.row},${token.pos.col})`);
        } else {
            console.log(token.toString());
        }
        token = lexer.parse();
    }
}
