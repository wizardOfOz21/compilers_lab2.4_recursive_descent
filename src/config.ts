import { LexemMap } from "./lexer";
import { arrayToRegexMap, objectToRegexMap } from "./regex_utils";

// Лексическая структура языка:
// - Лексические домены заданы регулярными выражениями
// - Приоритет доменов задан порядком их вхождения в объект lexemDefinition
// - Каждое ключевое слово принадлежит отдельному домену
// - Для keyWords название домена совпадает с ключевым словом, для объявления используется рефлексия
// - namedSpecialValues – именованные символьные лексемы

export const keyWords = [
    "NIL",
    "PACKED",
    "OF",
    "ARRAY",
    "FILE",
    "SET",
    "RECORD",
    "END",
    "TYPE",
    "VAR",
    "CASE"
];

// [ ] { } ( ) \ ^ $ . | ? * +
export const namedSpecialValues = {
    LPAREN: "\\(",
    RPAREN: "\\)",
    COMMA: ",",
    POINTS: "\\.\\.",
    CARET: "\\^",
    PLUS: "\\+",
    MINUS: "-",
    LBRACKET: "\\[",
    RBRACKET: "\\]",
    EQUAL: "=",
    COLON: ":",
    SEMICOLON: ";",
};

export const lexemDefinition: LexemMap = {
    ...objectToRegexMap(namedSpecialValues),
    ...arrayToRegexMap(keyWords),
    IDENTIFIER: /\p{L}\w*/u,
    UNSIGNED_NUMBER: /[0-9]+(\.[0-9]+)?([eE][-+]?[0-9]+)?/,
    STRING: /'[^\p{Cc}]*?'/u,
    
    COMMENT: /\{[^}]*\}/u,
    NEWLINE: /\n/,
    NEWLINE_R: /\r\n/,
    SPACE: /[^\S\r\n]+/,
};
