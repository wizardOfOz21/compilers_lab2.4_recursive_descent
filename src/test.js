import { controlSymbols, lexemDefinition, keyWords } from "./config";
import Lexer, { Token } from "./lexer";

test("Идентификатор распознается", () => {
    const data = "testCounter1";
    let lexer = new Lexer(data, lexemDefinition);
    let lexem = lexer.parse();
    expect(lexem.toString()).toMatchSnapshot();
});

test("Целые беззнаковые числа распознаются", () => {
    const data = "011010 123 456 1 0";
    let lexer = new Lexer(data, lexemDefinition);
    let token = lexer.parse();
    while (!token.isEof()) {
        expect(token.toString()).toMatchSnapshot();
        token = lexer.parse();
    }
});

test("Числа с плавающей точкой распознаются", () => {
    const data = "1E+11 42E-11 42E11 42.11E+11 42.11E-11 42.11E11";
    let lexer = new Lexer(data, lexemDefinition);
    let token = lexer.parse();
    while (!token.isEof()) {
        expect(token.toString()).toMatchSnapshot();
        token = lexer.parse();
    }
});

test("Конец строки распознается", () => {
    const data = "";
    let lexer = new Lexer(data, lexemDefinition);
    let lexem = lexer.parse();
    expect(lexem.toString()).toMatchSnapshot();
});

test("Столбцы считаются корректно", () => {
    const data = "  1234";
    let lexer = new Lexer(data, lexemDefinition);
    let lexem = lexer.parse();
    expect(lexem.toString()).toMatchSnapshot();
});

test("Строки считаются после \\n корректно", () => {
    const data = "  \n\n\n    1234";
    let lexer = new Lexer(data, lexemDefinition);
    let lexem = lexer.parse();
    expect(lexem.toString()).toMatchSnapshot();
});

test("Строки считаются после \\r\\n корректно", () => {
    const data = "  \r\n1234";
    let lexer = new Lexer(data, lexemDefinition);
    let lexem = lexer.parse();
    expect(lexem.toString()).toMatchSnapshot();
});

test("Строки распознаются", () => {
    const data = "'adsas' 'hello world!' ";
    let lexer = new Lexer(data, lexemDefinition);
    let token = lexer.parse();
    while (!token.isEof()) {
        expect(token.toString()).toMatchSnapshot();
        token = lexer.parse();
    }
});

test("Управляющие символы в строках не распознаются", () => {
    const data = "'\nsdfsd'";
    let lexer = new Lexer(data, lexemDefinition);
    let token = lexer.parse();
    while (!token.isEof()) {
        expect(token.toString()).toMatchSnapshot();
        token = lexer.parse();
    }
});

test("Cпециальные значения распознаются корректно", () => {
    const data = keyWords.join(' ');
    let lexer = new Lexer(data, lexemDefinition);
    let token = lexer.parse();
    while (!token.isEof()) {
        expect(token.toString()).toMatchSnapshot();
        token = lexer.parse();
    }
});

const namedSpecialValues = {
    LPAREN: "(",
    RPAREN: ")",
    COMMA: ",",
    POINTS: "..",
    CARET: "^",
    PLUS: "+",
    MINUS: "-",
    LBRACKET: "[",
    RBRACKET: "]",
    EQUAL: "=",
    COLON: ":",
    SEMICOLON: ";",
};

test("Именованные специальные значения распознаются корректно", () => {
    const data = Object.values(namedSpecialValues).join(' ');
    let lexer = new Lexer(data, lexemDefinition);
    let token = lexer.parse();
    while (!token.isEof()) {
        expect(token.toString()).toMatchSnapshot();
        token = lexer.parse();
    }
});

test("Распознается самая длинная лексема", () => {
    const data = "VARCHAR";
    let lexer = new Lexer(data, lexemDefinition);
    let token = lexer.parse();
    while (!token.isEof()) {
        expect(token.toString()).toMatchSnapshot();
        token = lexer.parse();
    }
});