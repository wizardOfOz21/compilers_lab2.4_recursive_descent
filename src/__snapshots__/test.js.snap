// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cпециальные значения распознаются корректно 1`] = `"NIL (1,1): NIL"`;

exports[`Cпециальные значения распознаются корректно 2`] = `"PACKED (1,5): PACKED"`;

exports[`Cпециальные значения распознаются корректно 3`] = `"OF (1,12): OF"`;

exports[`Cпециальные значения распознаются корректно 4`] = `"ARRAY (1,15): ARRAY"`;

exports[`Cпециальные значения распознаются корректно 5`] = `"FILE (1,21): FILE"`;

exports[`Cпециальные значения распознаются корректно 6`] = `"SET (1,26): SET"`;

exports[`Cпециальные значения распознаются корректно 7`] = `"RECORD (1,30): RECORD"`;

exports[`Cпециальные значения распознаются корректно 8`] = `"END (1,37): END"`;

exports[`Cпециальные значения распознаются корректно 9`] = `"TYPE (1,41): TYPE"`;

exports[`Cпециальные значения распознаются корректно 10`] = `"VAR (1,46): VAR"`;

exports[`Cпециальные значения распознаются корректно 11`] = `"CASE (1,50): CASE"`;

exports[`Идентификатор распознается 1`] = `"IDENTIFIER (1,1): testCounter1"`;

exports[`Именованные специальные значения распознаются корректно 1`] = `"LPAREN (1,1): ("`;

exports[`Именованные специальные значения распознаются корректно 2`] = `"RPAREN (1,3): )"`;

exports[`Именованные специальные значения распознаются корректно 3`] = `"COMMA (1,5): ,"`;

exports[`Именованные специальные значения распознаются корректно 4`] = `"POINTS (1,7): .."`;

exports[`Именованные специальные значения распознаются корректно 5`] = `"CARET (1,10): ^"`;

exports[`Именованные специальные значения распознаются корректно 6`] = `"PLUS (1,12): +"`;

exports[`Именованные специальные значения распознаются корректно 7`] = `"MINUS (1,14): -"`;

exports[`Именованные специальные значения распознаются корректно 8`] = `"LBRACKET (1,16): ["`;

exports[`Именованные специальные значения распознаются корректно 9`] = `"RBRACKET (1,18): ]"`;

exports[`Именованные специальные значения распознаются корректно 10`] = `"EQUAL (1,20): ="`;

exports[`Именованные специальные значения распознаются корректно 11`] = `"COLON (1,22): :"`;

exports[`Именованные специальные значения распознаются корректно 12`] = `"SEMICOLON (1,24): ;"`;

exports[`Конец строки распознается 1`] = `"EOF (-1,-1): "`;

exports[`Распознается самая длинная лексема 1`] = `"IDENTIFIER (1,1): VARCHAR"`;

exports[`Столбцы считаются корректно 1`] = `"UNSIGNED_NUMBER (1,3): 1234"`;

exports[`Строки распознаются 1`] = `"STRING (1,1): 'adsas'"`;

exports[`Строки распознаются 2`] = `"STRING (1,9): 'hello world!'"`;

exports[`Строки считаются после \\n корректно 1`] = `"UNSIGNED_NUMBER (4,5): 1234"`;

exports[`Строки считаются после \\r\\n корректно 1`] = `"UNSIGNED_NUMBER (2,1): 1234"`;

exports[`Управляющие символы в строках не распознаются 1`] = `"ERROR (1,1): "`;

exports[`Управляющие символы в строках не распознаются 2`] = `"IDENTIFIER (2,1): sdfsd"`;

exports[`Управляющие символы в строках не распознаются 3`] = `"ERROR (2,6): "`;

exports[`Целые беззнаковые числа распознаются 1`] = `"UNSIGNED_NUMBER (1,1): 011010"`;

exports[`Целые беззнаковые числа распознаются 2`] = `"UNSIGNED_NUMBER (1,8): 123"`;

exports[`Целые беззнаковые числа распознаются 3`] = `"UNSIGNED_NUMBER (1,12): 456"`;

exports[`Целые беззнаковые числа распознаются 4`] = `"UNSIGNED_NUMBER (1,16): 1"`;

exports[`Целые беззнаковые числа распознаются 5`] = `"UNSIGNED_NUMBER (1,18): 0"`;

exports[`Числа с плавающей точкой распознаются 1`] = `"UNSIGNED_NUMBER (1,1): 1E+11"`;

exports[`Числа с плавающей точкой распознаются 2`] = `"UNSIGNED_NUMBER (1,7): 42E-11"`;

exports[`Числа с плавающей точкой распознаются 3`] = `"UNSIGNED_NUMBER (1,14): 42E11"`;

exports[`Числа с плавающей точкой распознаются 4`] = `"UNSIGNED_NUMBER (1,20): 42.11E+11"`;

exports[`Числа с плавающей точкой распознаются 5`] = `"UNSIGNED_NUMBER (1,30): 42.11E-11"`;

exports[`Числа с плавающей точкой распознаются 6`] = `"UNSIGNED_NUMBER (1,40): 42.11E11"`;
